interpreter/decorators.go

```diff
@@ -76,47 +77,128 @@ func decDisableShortcircuits() InterpretableDecorator {
-// constant values and constructs a new aggregate value. Future improvements to this method will
-// also do the same for typed object creations and functions whose arguments are constant.
-func decFoldConstants() InterpretableDecorator {
+// decOptimize optimizes the program plan by looking for common evaluation patterns and
+// conditionally precomputating the result.
+// - build list and map values with constant elements.
+// - convert 'in' operations to set membership tests if possible.
+func decOptimize() InterpretableDecorator {
 	return func(i Interpretable) (Interpretable, error) {
 		switch i.(type) {
 		case *evalList:
-			l := i.(*evalList)
-			for _, elem := range l.elems {
-				_, isConst := elem.(*evalConst)
-				if !isConst {
-					return i, nil
-				}
-			}
-			val := l.Eval(EmptyActivation())
-			return &evalConst{
-				id:  l.id,
-				val: val,
-			}, nil
+			return maybeBuildListLiteral(i)
 		case *evalMap:
-			mp := i.(*evalMap)
-			for idx, key := range mp.keys {
-				_, isConst := key.(*evalConst)
-				if !isConst {
-					return i, nil
-				}
-				_, isConst = mp.vals[idx].(*evalConst)
-				if !isConst {
-					return i, nil
-				}
+			return maybeBuildMapLiteral(i)
+		case *evalBinary:
+			call := i.(*evalBinary)
+			if call.overload == overloads.InList {
+				return maybeOptimizeSetMembership(i)
 			}
-			val := mp.Eval(EmptyActivation())
-			return &evalConst{
-				id:  mp.id,
-				val: val,
-			}, nil
 		}
 		return i, nil
 	}
 }
 
+func maybeBuildListLiteral(i Interpretable) (Interpretable, error) {
+	l := i.(*evalList)
+	for _, elem := range l.elems {
+		_, isConst := elem.(*evalConst)
+		if !isConst {
+			return i, nil
+		}
+	}
+	val := l.Eval(EmptyActivation())
+	return &evalConst{
+		id:  l.id,
+		val: val,
+	}, nil
+}
+
+func maybeBuildMapLiteral(i Interpretable) (Interpretable, error) {
+	mp := i.(*evalMap)
+	for idx, key := range mp.keys {
+		_, isConst := key.(*evalConst)
+		if !isConst {
+			return i, nil
+		}
+		_, isConst = mp.vals[idx].(*evalConst)
+		if !isConst {
+			return i, nil
+		}
+	}
+	val := mp.Eval(EmptyActivation())
+	return &evalConst{
+		id:  mp.id,
+		val: val,
+	}, nil
+}
+
+func maybeOptimizeSetMembership(i Interpretable) (Interpretable, error) {
+	call := i.(*evalBinary)
+	if call.overload != overloads.InList {
+		return i, nil
+	}
+	l, isConst := call.rhs.(*evalConst)
+	if !isConst {
+		return i, nil
+	}
+	list := l.val.(traits.Lister)
+	if list.Size() == types.IntZero {
+		return &evalConst{
+			id:  call.id,
+			val: types.False,
+		}, nil
+	}
+	it := list.Iterator()
+	var typ ref.Type
+	valueSet := make(map[ref.Val]ref.Val)
+	for it.HasNext() == types.True {
+		elem := it.Next()
+		elemType := elem.Type()
+		if elemType != types.BoolType &&
+			elemType != types.BytesType &&
+			elemType != types.DoubleType &&
+			elemType != types.StringType &&
+			elemType != types.IntType &&
+			elemType != types.UintType {
```

Comment:
I see Timestamp and Duration are excluded, as they have the protobuf backing. I see no reason to generalize this type selection behavior for now, but maybe we could have a line explaining something like:

```// non-native types can't be optimized```